libexec/rtld-elf/rtld_c18n.c

```diff
@@ -1372,9 +1374,40 @@ tramp_check_sig(const struct tramp_header *found, const Obj_Entry *reqobj,
 		    C18N_SIG_FORMAT(data->sig),
 		    found->defobj->path, C18N_SIG_FORMAT(sig));
 	}
+
+	if (cheri_gettag(data->target)) {
+		found_target = atomic_load_explicit(&found->target,
+		    memory_order_relaxed);
+		do if (cheri_gettag(found_target)) {
+			rtld_require(
+			    cheri_equal_exact(found_target, data->target),
+			    "c18n: Incompatible capability metadata for "
+			    "function %s: found %#p, requested %#p\n",
+			    symname(found->defobj, found->symnum),
+			    found_target, data->target);
+			break;
+		} else if (atomic_compare_exchange_weak_explicit(&found->target,
+		    &found_target, data->target,
+		    memory_order_relaxed, memory_order_relaxed)) {
+			/*
+			 * Issue a data memory barrier so that the updated
+			 * trampoline target is visible to the trampoline when
+			 * it is run potentially in another thread.
+			 *
+			 * XXX Dapeng: The semantics of the fence being relied
```

Comment:
This makes me sad.  I guess you can't use an acquire barrier on the atomic_load above and then memory_order_release for the success case in the compare exchange?

Reply:
Even if I do that it won't fix the problem. The problem I'm trying to solve here is that after updating the target capability in the trampoline to a tagged one, the trampoline gets passed to another thread and gets called there, but still sees the untagged target in that thread.

The thing I really want is to make the memory store that passes the trampoline to another thread a store-release. Then it can be guaranteed that when the other threads loads the trampoline, it will see the updated target. But unfortunately I cannot control how the API user stores the trampoline.

Reply:
I've done some more research and it seems that although the C spec doesn't say so, virtually every online article claims that `atomic_thread_fence(memory_order_release)` prevents store-store ordering. So I guess I'd just tone down the comment a bit and leave the code as-is.